rllib/agents/trainer.py

```diff
@@ -1007,88 +1008,152 @@ def _sync_weights_to_workers(
+            unsquash_action: Should actions be unsquashed according to the
+                env's/Policy's action space? If None, use the value of
+                self.config["normalize_actions"].
+            clip_action: Should actions be clipped according to the
+                env's/Policy's action space? If None, use the value of
+                self.config["clip_actions"].
+
+        Keyword Args:
+            kwargs: forward compatibility placeholder
 
         Returns:
-            any: The computed action if full_fetch=False, or
-            tuple: The full output of policy.compute_actions() if
-                full_fetch=True or we have an RNN-based Policy.
+            The computed action if full_fetch=False, or a tuple of a) the
+                full output of policy.compute_actions() if full_fetch=True
+                or we have an RNN-based Policy.
 
         Raises:
             KeyError: If the `policy_id` cannot be found in this Trainer's
                 local worker.
         """
+        if clip_actions != DEPRECATED_VALUE:
+            deprecation_warning(
+                old="Trainer.compute_single_action(`clip_actions`=...)",
+                new="Trainer.compute_single_action(`clip_action`=...)",
+                error=False)
+            clip_action = clip_actions
+        if unsquash_actions != DEPRECATED_VALUE:
+            deprecation_warning(
+                old="Trainer.compute_single_action(`unsquash_actions`=...)",
+                new="Trainer.compute_single_action(`unsquash_action`=...)",
+                error=False)
+            unsquash_action = unsquash_actions
+
+        # User provided an input-dict: Assert that `obs`, `prev_a|r`, `state`
+        # are all None.
+        err_msg = "Provide either `input_dict` OR [`observation`, ...] as " \
+                  "args to Trainer.compute_single_action!"
+        if input_dict is not None:
+            assert observation is None and prev_action is None and \
+                   prev_reward is None and state is None, err_msg
+            observation = input_dict[SampleBatch.OBS]
+        else:
+            assert observation is not None, err_msg
+
+        # Get the policy to compute the action for (in the multi-agent case,
+        # Trainer may hold >1 policies).
         policy = self.get_policy(policy_id)
         if policy is None:
             raise KeyError(
                 f"PolicyID '{policy_id}' not found in PolicyMap of the "
                 f"Trainer's local worker!")
-
         local_worker = self.workers.local_worker()
 
-        if state is None:
-            state = []
-
         # Check the preprocessor and preprocess, if necessary.
         pp = local_worker.preprocessors[policy_id]
         if pp and type(pp).__name__ != "NoPreprocessor":
             observation = pp.transform(observation)
-        filtered_observation = local_worker.filters[policy_id](
+        observation = local_worker.filters[policy_id](
```

Comment:
this line feels really confusing, can you add a comment to explain the rationale a little bit?
above when we check input_dict validity, we set observation to input_dict[SampleBatch.OBS].
and then here, we will overwrite that with observation, which is local_worker.filters[policy_id].